### ISSUE TITLE
Travis command line tool does not respect https_proxy.

### ISSUE THREAD

**Issue by chmduquesne** [travis-ci/travis.rb#262]
My workplace modifies our https traffic with a Bluecoat SSL proxy. If you don't know what this is, it is a tool from hell that basically intercepts employees' https traffic and substitutes certificates, essentially performing a MITM attack and making literally every tool that requires https complain. One of these tools is travis.

For programs that are really security sensitive, such as gem, git or pip, we have an https proxy that does not perform such an attack. We use this proxy by exporting the variable https_proxy in our shell. The direct consequence is that the tools we are using need to support this variable. When using `travis login`, I am encountering the same error regardless the fact that the https_proxy variable is set or not. Here is the result of `travis report`:

```
System
Ruby:                     Ruby 2.1.5-p273
Operating System:         Debian 8.0
RubyGems:                 RubyGems 2.2.2

CLI
Version:                  1.7.5
Plugins:                  none
Auto-Completion:          yes
Last Version Check:       2015-01-21 09:22:49 +0100

Session
API Endpoint:             https://api.travis-ci.org/
Logged In:                no
Verify SSL:               yes
Enterprise:               no

Endpoints
org:                      https://api.travis-ci.org/ (current)

Last Exception
An error occurred running `travis login`:
    TypeError: no implicit conversion of nil into String
        from /usr/lib/ruby/2.1.0/json/common.rb:155:in `initialize'
        from /usr/lib/ruby/2.1.0/json/common.rb:155:in `new'
        from /usr/lib/ruby/2.1.0/json/common.rb:155:in `parse'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/tools/github.rb:248:in `gh_error'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/tools/github.rb:216:in `rescue in basic_auth'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/tools/github.rb:208:in `basic_auth'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/tools/github.rb:221:in `login'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/tools/github.rb:80:in `possible_tokens'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/tools/github.rb:49:in `each_token'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/tools/github.rb:36:in `with_token'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/cli/login.rb:31:in `login'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/cli/login.rb:40:in `run'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/cli/command.rb:198:in `execute'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/lib/travis/cli.rb:64:in `run'
        from /var/lib/gems/2.1.0/gems/travis-1.7.5/bin/travis:18:in `<top (required)>'
        from /usr/local/bin/travis:23:in `load'
        from /usr/local/bin/travis:23:in `<main>'


For issues with the command line tool, please visit https://github.com/travis-ci/travis.rb/issues.
For Travis CI in general, go to https://github.com/travis-ci/travis-ci/issues or email support@travis-ci.com.
```

I understand that such issues are laughable and that the fix should be on the network side, but this is preventing me to use travis fully. I am trying to convince my management that our company should consider using it (making a presentation soon) and let us say that this issue will not help during the demo.

**Comment by FloFra** [comment]
travis.rb uses faraday as its http library which supports the use of a proxy. The problem right now is that travis.rb does not allow you to access these settings.

**Comment by adriens** [comment]
Any chance to see that running soon ? It's really blocking in corporate environments.

Thank you in advance for your help.

**Comment by thojansen** [comment]
I can confirm that the CLI is not working properly behind web proxies. For us, it might even be a missing no_proxy support.
https://github.com/lostisland/faraday/pull/471

**Comment by gabel** [comment]
`no_proxy` support was finally merged into faraday master via https://github.com/lostisland/faraday/pull/658 - so there may be an release someday and maybe travis.rb can update to that faraday version

**Comment by BorisNaguet** [comment]
Hello,
Any news on that issue?
I have to remove "no_proxy" env variable before using travis CLI.
This is a terrible user experience when 1st using the tool.

From what I understand, Faraday 0.12 is released

**Comment by Finkes** [comment]
any updates on this? It's been quite a while

**Comment by pboyd04** [comment]
This is still an issue for corporate use of this tool

**Comment by frostming** [comment]
Ping, any workaround here to get me through?

**Comment by garlicsauce** [comment]
up, any workaround?

**Comment by tobiasmasch** [comment]
Is there any solution available?

**Comment by BanzaiMan** [comment]
Is this still an issue? https://lostisland.github.io/faraday/usage/customize#proxy leads me to believe that Faraday supports the use of `http_proxy` and others to configure the HTTP connexion.

**Comment by BanzaiMan** [comment]
This should be fixed at this point. Failing to connect to the proxy can lead to `NameError`, with the current 1.9.1:

```
uninitialized constant Faraday::Error::ClientError
```

which is fixed by #751.

After the fix, it should look like this:

```sh-session
$ ./bin/travis --version
1.9.1
$ env | grep _proxy
$ ./bin/travis history --limit 1 --pro
#1233 started:   master (PR #751) Fix Faraday error constant (for v1.x) and typo
$ export https_proxy=https://proxy.example.com
$ ./bin/travis history --limit 1 --pro
SSL error: could not verify peer
$ ./bin/travis history --limit 1 --pro --debug-http
Failed to open TCP connection to proxy.example.com:443 (getaddrinfo: nodename nor servname provided, or not known)
for a full error report, run travis report --pro
```

I'm closing this, as the issue at hand is fixed.
